.github/workflows/build-and-test.yml

```diff
@@ -210,3 +210,95 @@ jobs:
             exit 1
           fi
           juju exec --unit microceph/0 -- microceph status
+
+  juju-upgrade-test:
+    needs:
+      - lint
+      - unit-test
+      - build
+    name: Juju upgrade test
+    runs-on: [self-hosted, xlarge]
+    steps:
+
+      - name: Download charm
+        uses: actions/download-artifact@v3
+        with:
+          name: charms
+          path: ~/artifacts/
+
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Setup LXD
+        uses: canonical/setup-lxd@v0.1.1
+
+      - name: Install dependencies
+        run: |
+          date
+          sudo snap install juju
+          mkdir -p ~/.local/share/juju
```

Comment:
Just a nit: But portions of these tests could be moved to the `ci_helpers.sh` script as they appear often in our tests.